paddle/operators/clip_op.cu

```diff
@@ -0,0 +1,72 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#include "paddle/operators/clip_op.h"
+
+#define CUDA_1D_KERNEL_LOOP(i, n)                            \
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; \
+       i += blockDim.x * gridDim.x)
+
+namespace paddle {
+namespace operators {
+
+using framework::LoDTensor;
+
+template <typename T>
+__global__ void ClipGradientKernel(const int N, const T min, const T max,
+                                   const T* Y, const T* dY, T* dX) {
+  CUDA_1D_KERNEL_LOOP(i, N) {
+    if (Y[i] > min && Y[i] < max) {
+      dX[i] = dY[i];
+    } else {
+      dX[i] = 0;
+    }
+  }
+}
+
+template <typename T>
+class ClipGradientOpCUDAKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto max = context.Attr<float>("max");
+    auto min = context.Attr<float>("min");
+    auto* d_out = context.Input<LoDTensor>(framework::GradVarName("Out"));
+    auto* d_x = context.Output<LoDTensor>(framework::GradVarName("X"));
+    auto* x = context.Input<LoDTensor>("X");
+    auto dims = d_x->dims();
+    int64_t count = d_out->numel();
+    auto d_x_data = d_x->mutable_data<T>(context.GetPlace());
```

Comment:
Why not just `d_x->mutable_data<T>(context.GetPlace())` and keep using `d_x`?

Reply:
`d_x_data` is needed by [ClipGradientKernel](https://github.com/PaddlePaddle/Paddle/pull/3937/files#diff-a962b1a6c083e76106209b8b60f69dafR61)

Reply:
I mean `d_x_data` and `d_x` they are totally same. They point to the same memory.